packages/wallet/ui/src/bridge-dapp.jsx

```diff
@@ -0,0 +1,180 @@
+// @ts-check
+import '@endo/eventual-send/shim';
+import './lockdown.js';
+
+import React, { useEffect } from 'react';
+import ReactDOM from 'react-dom';
+import { BridgeProtocol } from '@agoric/web-components';
+import { addOffer } from './store/Offers.js';
+import { loadDapp, upsertDapp, watchDapps } from './store/Dapps.js';
+
+Error.stackTraceLimit = Infinity;
+
+/**
+ * Sends a message to the dapp in the parent window.
+ *
+ * @param {*} payload
+ */
+const sendMessage = payload => window.parent.postMessage(payload, '*');
+
+/**
+ * Ensures this bridge is being loaded inside another window.
+ */
+const checkParentWindow = () => {
+  const me = window;
+  const { parent } = window;
+  if (me === parent) {
+    throw Error('window.parent === parent!!!');
+  }
+};
+
+/**
+ * Requests a dapp at a given origin to connect to the wallet UI. This will
+ * prompt the user to accept the dapp in the wallet and allow it to propose
+ * offers if accepted.
+ *
+ * @param {string} origin
+ * @param {string} chainId
+ * @param {string} address
+ * @param {string} proposedPetname
+ */
+const requestDappConnection = (origin, chainId, address, proposedPetname) => {
+  const dapp = loadDapp(chainId, address, origin);
+  if (dapp) {
+    return;
+  }
+  upsertDapp(chainId, address, { origin, petname: proposedPetname });
```

Comment:
Surprising to me that we up-sert instead of insert even when the dapp is known to not be present. I suppose because there's no `insert`. And we need to skip the upsert if the dapp is present because we don't want to update the petname. It would help to have this explained in the comments. At the least, have `proposedPetname` jsdoc indicate that it will only be used if the dapp identified by the other three args isn't present.

Reply:
Added a jsdoc description